### PR TITLE
feat(mcp): fold conversation-history recall into chat_ask

### PR DESCRIPTION
**Pull request by teng-lin**
## What

Folds chat **history recall** into the existing `chat_ask` MCP tool via an optional `history: int` param, instead of adding a new `chat_history` tool — the MCP manifest is at its **28/28 ceiling**, so this stays ceiling-neutral (the design call from the [2026-06-29 MCP gap review](../blob/main/docs/notes/mcp-gap-review-2026-06-29.md), which ranks chat-history recall **Critical**).

`async chat` (start/poll) was **deliberately not built**: chat is a single RPC round-trip (no backend job/poll like research/artifacts), so it would require stateful in-process task tracking that breaks the server's stateless design — revisit only if a real chat timeout is reported.

## Behavior

- `question` is now **optional**; new `history: int = 0` param.
- `history > 0` → response carries up to N prior `{question, answer}` **pairs** (oldest-first), fetched **before** the ask so the new turn isn't double-reported.
- **Recall-only** (omit `question`, `history > 0`) → returns `history` and echoes the resolved `conversation_id`; no ask.
- Empty call (no question, no history) → `ValidationError`.
- Reuses existing `client.chat.get_history`; **no RPC changes**.

## Correctness details (from review)

- `limit` counts individual role-rows (~2/pair) → **doubled** so `history=N` ≈ N pairs.
- When recall + a question both target the most-recent conversation, the id is **resolved once and pinned** to both reads/writes (no diverging awaits); the pure-ask hot path pays no extra round-trip.
- Whitespace-only question → treated as recall path, never a blank ask.

## Tests

7 new `test_chat.py` cases: recall-only (+echo, doubled limit), empty-conversation recall, question+history pinning, zero-history skip, whitespace rejection, missing-arg rejection. Full unit suite green (9200 passed); mypy + ruff clean; manifest ceiling holds.

Docs: `mcp-guide.md` signature table + `architecture.md` per-file index updated.

🤖 Generated with [Claude Code](https://claude.com/claude-code)

https://claude.ai/code/session_012ZUzFEAC3XBxwjYQ435QFQ

<a href="https://cubic.dev/pr/teng-lin/notebooklm-py/pull/1714?utm_source=github" target="_blank" rel="noopener noreferrer" data-no-image-dialog="true"><picture><source media="(prefers-color-scheme: dark)" srcset="https://www.cubic.dev/buttons/review-in-cubic-dark.svg"><source media="(prefers-color-scheme: light)" srcset="https://www.cubic.dev/buttons/review-in-cubic-light.svg">Review in cubic</picture></a>

## Summary by CodeRabbit

* **New Features**
  * `chat_ask` now supports recalling prior question-and-answer turns via an optional `history` parameter, including returning the resolved conversation ID.
* **Bug Fixes**
  * Improved input validation for empty/whitespace questions; tool returns history-only results without triggering a new chat answer when appropriate.
* **Documentation**
  * Updated the architecture and tool-reference docs to reflect the new `question`/`history` behavior.
* **Tests**
  * Added unit tests covering recall paths, edge cases, and validation failures.

### CODE REVIEW

**Comment by teng-lin** [comment]
@claude review

File: src/notebooklm/mcp/tools/chat.py [hunk 1]
```diff
@@ -63,36 +66,74 @@ async def chat_ask(
...
         """
         client = get_client(ctx)
         with mcp_errors():
+            # A whitespace-only question counts as "no question" (recall path), so
+            # a blank string can't slip past the guard into client.chat.ask.
+            question = question.strip()
+            if not question and history <= 0:
+                raise ValidationError(
+                    "Provide a question to ask, or history>0 to recall prior turns."
+                )
```
[the remark this answers is not in the document]
Comment (teng-lin) [reply src/notebooklm/mcp/tools/chat.py]: Addressed in db375566: added `if history < 0: raise ValidationError("history must be >= 0.")` before the empty-call guard, so a negative `history` (even with a question) now fails as invalid input rather than silently degrading to a plain ask. Added `test_chat_ask_negative_history_rejected`.

[the remark this answers is not in the document]
Comment (teng-lin) [reply src/notebooklm/mcp/tools/chat.py]: Addressed in db375566: negative `history` now raises `ValidationError` explicitly (`history must be >= 0.`), guarded ahead of the existing empty-call check. Covered by `test_chat_ask_negative_history_rejected`.

**Comment by teng-lin** [comment]
Addressed @claude review in db375566:
- **Redundant `get_conversation_id`**: when no conversation exists, `chat_ask` now short-circuits to `history: []` instead of calling `get_history` with `conversation_id=None` (which re-resolved the absent id). The non-empty/explicit-id paths are unchanged.
- **Test coverage**: added `test_chat_ask_history_explicit_conversation_id_skips_resolve` (asserts `get_conversation_id.assert_not_called()` when an explicit `conversation_id` is supplied) and `test_chat_ask_recall_only_empty_conversation` now asserts `get_history` is not awaited.
